lib/filewatch/watched_file.rb

```diff
@@ -0,0 +1,180 @@
+require "filewatch/buftok"
+
+module FileWatch
+  class WatchedFile
+    def self.new_initial(path, inode, stat = nil)
+      new(path, inode, stat, true)
+    end
+
+    def self.new_ongoing(path, inode, stat = nil)
+      new(path, inode, stat, false)
+    end
+
+    attr_reader :size, :inode, :state, :file, :buffer, :state_history
+    attr_reader :path, :filestat, :ignored_size, :accessed_at
+    attr_accessor :close_older, :ignore_older, :delimiter
+
+    def delimiter
+      @delimiter
+    end
+
+    def initialize(path, inode, stat, initial)
+      @path = path
+      @ignored_size = @size = 0
+      @inode = inode
+      @initial = initial
+      @state_history = []
+      @state = :watched
+      if (@filestat = stat).nil? && !@inode.nil?
+        restat
+      end
+      set_accessed_at
+    end
+
+    def set_accessed_at
+      @accessed_at = Time.now.to_i
+    end
+
+    def initial?
+      @initial
+    end
+
+    def size_changed?
+      filestat.size != size
+    end
+
+    def inode_changed?(value)
+      self.inode != value
+    end
+
+    def file_add_opened(rubyfile)
+      @file = rubyfile
+      @buffer = FileWatch::BufferedTokenizer.new(delimiter || "\n")
+    end
+
+    def file_close
+      return if @file.nil? || @file.closed?
+      @file.close
+      @file = nil
+    end
+
+    def file_seek(amount, whence = IO::SEEK_SET)
+      @file.sysseek(amount, whence)
+    end
+
+    def file_read(amount)
+      set_accessed_at
+      @file.sysread(amount)
+    end
+
+    def file_open?
+      !@file.nil? && !@file.closed?
+    end
+
+    def update_read_size(total_bytes_read)
+      return if total_bytes_read.nil?
+      @size = total_bytes_read
+    end
+
+    def buffer_extract(data)
+      @buffer.extract(data)
+    end
+
+    def update_inode(_inode)
+      @inode = _inode
+    end
+
+    def update_size
+      @size = @filestat.size
+    end
+
+    def activate
+      archive_state
+      @state = :active
+    end
+
+    def ignore
+      archive_state
+      @state = :ignored
+      @ignored_size = @size = @filestat.size
```

Comment:
shouldn't ignore be based on the `@size` snapshot? from what I understand it represents how far a file has been read? what is the added benefit of this second `@ignored_size`?